src/branch_and_bound.jl

```diff
@@ -0,0 +1,251 @@
+import Base: copy, eltype, iterate, IteratorSize
+import Base: getindex, setindex!, delete!
+
+export BBSearch, BreadthFirstBBSearch, DepthFirstBBSearch
+export data
+export copy, eltype, iterate, IteratorSize
+
+abstract type AbstractBBNode end
+
+"""
+    BBNode <: AbstractBBNode
+
+Intermediate node of a `BBTree`. Does not contain any data by itself,
+only redirect toward its children.
+"""
+struct BBNode <: AbstractBBNode
+    parent::Int
+    children::Vector{Int}
+end
+
+"""
+    BBLeaf{DATA} <: AbstractBBLeaf
+
+Leaf node of a `BBTree`. Contains both data and a status indicating if it
+will be further processed by the branch and bound search.
+"""
+struct BBLeaf{DATA} <: AbstractBBNode
+    data::DATA
+    parent::Int
+    status::Symbol
+end
+
+function BBLeaf(data::DATA, parent::Int) where {DATA}
+    BBLeaf{DATA}(data, parent, :working)
+end
+
+function BBNode(leaf::BBLeaf, child1::Int, child2::Int)
+    BBNode(parent_id(leaf), Int[child1, child2])
+end
+
+"""
+    BBTree{DATA}
+
+Tree storing the data used and produced by a branch and bound search in a
+structured way.
+
+Node can be accessed using their index using the bracket syntax `wt[node_id]`.
+Support the iterator interface. The element yielded by the iteration are
+tuples `(node_id, lvl)` where `lvl` is the depth of the node in the tree.
+"""
+struct BBTree{DATA}
+    nodes::Dict{Int, Union{BBNode, BBLeaf{DATA}}}
```

Comment:
Is there a way of avoiding this `Union`?

Reply:
It is possible to have two dict: one for the branchings (`BBNode`) and one for the leaves (`BBLeaf{DATA}`).